lib/FusionInventory/Agent/Tools/Linux.pm

```diff
@@ -385,54 +385,69 @@ sub _readLinkFromSysFs {
 sub getInfoFromSmartctl {
     my (%params) = @_;
 
-    my $handle = getFileHandle(
-        %params,
-        command => $params{device} ? "smartctl -i $params{device}" : undef,
-    );
-    return unless $handle;
+    $params{command} = $params{device} ? "smartctl -i $params{device} " : undef;
+    $params{command} .= $params{extra} if (defined $params{command} and defined $params{extra});
 
     my $info = {
         TYPE        => 'disk',
         DESCRIPTION => 'SATA',
     };
 
-    while (my $line = <$handle>) {
-        if ($line =~ /^Vendor: +(\S+)/i) {
-            $info->{MANUFACTURER} = getCanonicalManufacturer($1);
-            next;
+    my $attrs = {
+        MANUFACTURER => {
+            src  => ['vendor', 'model family', 'add. product id', 'device model', 'product'],
+            func => \&getCanonicalManufacturer,
+            args => undef
+        },
+        MODEL => {
+            src => ['product', 'device model', 'model family']
+        },
+        FIRMWARE => {
+            src => ['revision', 'firmware version']
+        },
+        DISKSIZE => {
+            src  => ['user capacity'],
+            func => \&getCanonicalSize,
+            args => [1024]
+        },
+        DESCRIPTION => {
+            src => ['transport protocol']
+        },
+        TYPE => {
+            src => ['device type']
+        },
+        SERIALNUMBER => {
+            src => ['serial number']
         }
+    };
 
-        if ($line =~ /^Product: +(\S+)/i) {
-            $info->{MODEL} = $1;
-            next;
-        }
+    my $regexp = {
+        __default       => qr/^(\w+)/,
+        __smartctl      => qr/^([^:]+?)\s*:\s*(.+)\s*$/,
+        'user capacity' => qr/([\d\.\,\s]+(?:\w+)?)/,
+        'device model'  => qr/([\w\s]+)/,
+    };
 
-        if ($line =~ /^Revision: +(\S+)/i) {
-            $info->{FIRMWARE} = $1;
-            next;
-        }
+    my %smartctl;
 
-        if ($line =~ /^User Capacity: +(\S.+\S)/i) {
-            $info->{DISKSIZE} = getCanonicalSize($1, 1024);
-            next;
-        }
+    for my $line (getAllLines(%params)) {
```

Comment:
```suggestion
    for my $line (@lines) {
```

Reply:
Just to follow my first code suggestion.

Reply:
Updated the code.